File: C++/Entity.cpp
#include "Entity.h"
#include <iostream>

Entity& Entity::takeDamage(int damageTaken) {
	damageTaken -= Defense;
	if (damageTaken > 0) {
		this->Health -= damageTaken;
	}
	return *this;
}

Entity& Entity::operator=(const Entity& other) {
	if (this == &other) {
		return *this;
	}
	Name = other.Name;
	Health = other.Health;
	Defense = other.Defense;
	Attack = other.Attack;
	Level = other.Level;
	x = other.x;
	y = other.y;
	Symbol = other.Symbol;
	inventory = other.inventory;
	return *this;
}

char Entity::getChar() {
	return Symbol;
}

std::string Entity::getName() {
	return Name;
}

int Entity::getHealth() {
	return Health;
}

std::pair<int, int> Entity::getPosition() const {
	return { x, y };
}

void Entity::setPosition(int newX, int newY) {
	x = newX;
	y = newY;
}

int Entity::totalEntities = 0;

Entity::Entity() {
	totalEntities++;
}

Entity::Entity(std::string Name, int Health, int Defense, int Attack, int Level, char Symbol)
	: Name(Name), Health(Health), Defense(Defense), Attack(Attack), Level(Level), Symbol(Symbol) {
	totalEntities++;
}

int Entity::getTotalEntities() {
	return totalEntities;
}

std::ostream& operator<<(std::ostream& os, const Entity& entity) {
	os << "Name: " << entity.Name << "\n"
		<< "Health: " << entity.Health << "\n"
		<< "Defense: " << entity.Defense << "\n"
		<< "Attack: " << entity.Attack << "\n"
		<< "Level: " << entity.Level << "\n"
		<< "Position: (" << entity.x << ", " << entity.y << ")\n"
		<< "Symbol: " << entity.Symbol << "\n";
	return os;
}

File: Java/Entity.java
package Java;

import java.util.ArrayList;
import java.util.List;

public abstract class Entity {
    protected String name;
    protected char symbol;
    protected int health, defense, attack, level;
    protected int x, y;
    public Inventory<Item> inventory;
    private static int totalEntities = 0;

    public Entity() {
        totalEntities++;
        inventory = new Inventory<Item>();
    }

    public Entity(String name, int health, int defense, int attack, int level, char symbol) {
        this.name = name;
        this.health = health;
        this.defense = defense;
        this.attack = attack;
        this.level = level;
        this.symbol = symbol;
        totalEntities++;
        inventory = new Inventory<Item>();
    }

    public static int getTotalEntities() {
        return totalEntities;
    }

    protected Entity takeDamage(int damageTaken) {
        health -= DamageCalculator.calculateDamage(damageTaken, defense);
        return this;
    }

    public char getChar() {
        return symbol;
    }

    public int getHealth() {
        return health;
    }

    public String getName() {
        return name;
    }

    public int[] getPosition() {
        return new int[]{x, y};
    }

    public void setPosition(int newX, int newY) {
        x = newX;
        y = newY;
    }

    @Override
    public String toString() {
        return "Name: " + name + "\n" +
               "Health: " + health + "\n" +
               "Defense: " + defense + "\n" +
               "Level: " + level + "\n" +
               "Position: (" + x + ", " + y + ")\n" +
               "Symbol: " + symbol + "\n";
    }

    protected abstract void onLevelUp();
}

class DamageCalculator {
    public static int calculateDamage(int baseDamage, int defense) {
        int damage = baseDamage - defense;
        return Math.max(damage, 0); // Урон не может быть отрицательным
    }
}